libsolidity/codegen/ir/IRGenerator.cpp

```diff
@@ -239,64 +271,166 @@ string IRGenerator::generateInitialAssignment(VariableDeclaration const& _varDec
+					*(arguments->at(i)),
+					*(baseConstructor->parameters()[i]->type())
+			).commaSeparatedList();
 
-	using boost::adaptors::reverse;
+		return params;
+	};
 
-	ostringstream out;
+	if (FunctionDefinition const* constructor = _contract.constructor())
+		for (auto const& modifier: constructor->modifiers())
+		{
+			FunctionDefinition const* baseConstructor = dynamic_cast<ContractDefinition const*>(
+					modifier->name()->annotation().referencedDeclaration
+			)->constructor();
 
-	FunctionDefinition const* constructor = _contract.constructor();
-	if (constructor && !constructor->isPayable())
-		out << callValueCheck();
+			if (baseConstructor && modifier->arguments() && !modifier->arguments()->empty())
+				m_baseConstructorParameters[baseConstructor] = listParamNames(baseConstructor, modifier->arguments());
+		}
 
-	for (ContractDefinition const* contract: reverse(_contract.annotation().linearizedBaseContracts))
+	for (ASTPointer<InheritanceSpecifier> const& base: _contract.baseContracts())
 	{
-		out <<
-			"\n// Begin state variable initialization for contract \"" <<
-			contract->name() <<
-			"\" (" <<
-			contract->stateVariables().size() <<
-			" variables)\n";
-
-		IRGeneratorForStatements generator{m_context, m_utils};
-		for (VariableDeclaration const* variable: contract->stateVariables())
-			if (!variable->isConstant() && !variable->immutable())
-				generator.initializeStateVar(*variable);
-		out << generator.code();
-
-		out << "// End state variable initialization for contract \"" << contract->name() << "\".\n";
+		FunctionDefinition const* baseConstructor = dynamic_cast<ContractDefinition const*>(
+				base->name().annotation().referencedDeclaration
+		)->constructor();
+
+		if (baseConstructor && base->arguments() && !base->arguments()->empty())
+			m_baseConstructorParameters[baseConstructor] = listParamNames(baseConstructor, base->arguments());
 	}
 
-	if (constructor)
-	{
-		ABIFunctions abiFunctions(m_evmVersion, m_context.revertStrings(), m_context.functionCollector());
-		unsigned paramVars = make_shared<TupleType>(constructor->functionType(false)->parameterTypes())->sizeOnStack();
+	return generator.code();
+}
+
+string IRGenerator::initStateVariables(ContractDefinition const& _contract)
+{
+	IRGeneratorForStatements generator{m_context, m_utils};
+	for (VariableDeclaration const* variable: _contract.stateVariables())
+		if (!variable->isConstant() && !variable->immutable())
+			generator.initializeStateVar(*variable);
+
+	Whiskers initStateVarsCode(R"X(
+		// Begin state variable initialization for contract <contractName> (<numVariables>)
+		<stateVariableCode>
+		// End state variable initialization for contract "<contractName>".
+	)X");
+	initStateVarsCode("contractName", _contract.name());
+	initStateVarsCode("numVariables", toString(_contract.stateVariables().size()));
+	initStateVarsCode("stateVariableCode", generator.code());
+
+	return initStateVarsCode.render();
```

Comment:
```suggestion
	return Whiskers(R"(
		// Begin state variable initialization for contract <contractName> (<numVariables>)
		<stateVariableCode>
		// End state variable initialization for contract "<contractName>".
	)")
	("contractName", _contract.name())
	("numVariables", toString(_contract.stateVariables().size()))
	("stateVariableCode", generator.code()).render();
```
?

Also I'm not *entirely* sure the additional comments justify the use of the full-blown template...
I'd be fine with just ``return generator.code();``... but the comments might indeed be helpful, so I'm fine either way.

Reply:
I added it as it was before like this, but I think I can remove as it doesn't mean a lot in context